### ISSUE TITLE
collect() and dependencies current issues

### ISSUE THREAD

**Issue by ethauvin** [cbeust/kobalt#392]
Assuming:

```
    dependencies {
        compile("org.apache.velocity:velocity:1.7")
    }
```

which resolves to:
```
╟ commons-collections:commons-collections:jar:3.2.1
╟ commons-lang:commons-lang:jar:2.4
╙ oro:oro:jar:2.0.8 (optional)
```

`collect(compileDependencies) ` will return `velocity`, `commons-collections` and `commons-lang` as expected.

Change to:

```
    dependencies {
        compile("org.apache.velocity:velocity:jar:1.7")
    }
```

`collect(compileDependencies)` will still return all three dependencies, not just `velocity`.

**Comment by ethauvin** [comment]
`collect(compileDependencies)` returns `List<File>`, which works in most case.

Assuming:

```
    dependencies {
        compile("org.apache.velocity:velocity:1.7")
        compile("example.com:foo:bar:1.0")
    }
````

How can I have `velocity` copied to my install `target` directory, but not `foo.bar`?

```
install {
        target = deploy
        collect(compileDependencies).forEach {
            if (it.absolutePath.contains("velocity")) {
                copy(from(it.absolutePath), to("$target/lib"))
            }
        }
}
```

Pretty weak, IMHO.

Is it viable to make `collect()` return both `files` and Maven `coordinates`?

**Comment by cbeust** [comment]
Sure but how would that help? You'd still need an if here.


-- 
Cédric

**Comment by ethauvin** [comment]
True, but it is a very weak `if`, because I have no clues as to what the file name is.

Take a scenario when multiple packages use a different version of 'commons-lang`, but I would only want the most recent version to be in my deployment directory? Unless I hard-code the filename, there's no way to do it. On the other hand, if I can compare against my previously specified coordinates, I'm good.

**Comment by ethauvin** [comment]
As discussed, `compileAptDependencies` should be implemented for use in `collect()`

**Comment by cbeust** [comment]
So should I add a `collectIds()` directive or have `collect()` return both the file name and id?

**Comment by ethauvin** [comment]
I'm not wild about `collectIds()`.

`collect()` could return both, or you could use a flag parameter, `collect(dependencies: List<IClasspathDependency>, mavenIds: Boolean = false)` or something like that.

**Comment by ethauvin** [comment]
I take that back, `collect()` returning both, is best. It'll allow for one call to `collect()` to check for multiple options, e.g.:

```
collect(compileDependencies).forEach {
    if (it.file.abolutePath.equals("blah")) {
        ...
    } else if (it.id.equals("foo") {
        ...
    }
}

**Comment by ethauvin** [comment]
Any comments on https://github.com/cbeust/kobalt/issues/392#issue-220091021?

That's a big one for me, and I think that's a caching issue.

**Comment by ethauvin** [comment]
Let me expand on https://github.com/cbeust/kobalt/issues/392#issue-220091021

I'm using annotator, which requires both `apt(annotatorJar)` and `compile(annotatorJar)`, like most do.

The problem is when I'm using `collect(compileDependencies)`, I'm getting the annotator jar and it's dependencies, but I don't really need them to be deployed to the `install` target.

Gradle addresses that issue by providing a `compileOnly 'annotatorJar'` directive. Meaning that when collecting the compile dependencies, no annotator dependencies will be included.

**Comment by cbeust** [comment]
Fixed in 1.0.55.
